CARDS-2618: Replace Livetable with MaterialReactTable in administration screens

veronikaslc opened the pull request:
[CARDS-2618](https://phenotips.atlassian.net/browse/CARDS-2618): Replace Livetable with MaterialReactTable in administration screens

Affected Administration components:
- Subject types
- Clinics
- Statistics
- Questionnaires

Dataflow change:
On the Subject types page the newly created subjet type immediately appears in the table after successful creation

[CARDS-2618]: https://phenotips.atlassian.net/browse/CARDS-2618?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

veronikaslc committed:
CARDS-2618: Replace Livetable with MaterialReactTable in administration screens

veronikaslc committed:
CARDS-2618: Replace Livetable with MaterialReactTable in administration screens
Add data to the Subject Types table on successfull creation

marta- commented:
The search function seems to have some issues.

I started cards with `-P prems`.
* In the `Clinics` admin screen I searched for "inpatient" and all the inpatient clinics were retained, PLUS "Canadian Patient Experience". After a couple of more searches I concluded that if you search for a string `abc`, it actually matches `*a*b*c*`. I'm sure it's meant to be a cool feature but the extra search results feel more like noise, especially since there's no highlighting for inexact matches.
* In `Questionnaires` the search doesn't seem to work at all.

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/EditButton.jsx:
```diff
@@ -28,17 +28,20 @@ import { Link } from 'react-router-dom';
-        <IconButton className={className} size={size}>
+  const { entryPath, entryType, size, className, admin, onClick, useEditDialog } = props;
+
+  let innerButton = <Tooltip title={entryType ? "Edit " + entryType : "Edit"}>
```
why was `toLowerCase()` dropped?

veronikaslc replied:
To preserve the [original styling](https://github.com/data-team-uhn/cards/blob/dev/modules/data-entry/src/main/frontend/src/dataHomepage/SubjectTypes.jsx#L42)

marta- replied:
But by preserving this, which is the odd one out, we alter the tooltips of many other occurrences of the edit button. Sentence case makes more sense, please put `toLowerCase()` back.

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/EditButton.jsx:
```diff
@@ -28,17 +28,20 @@ import { Link } from 'react-router-dom';
         </IconButton>
       </Tooltip>
+
+  return useEditDialog ? innerButton : (
```
`useEditDialog` seems unnecessary. If `onClick` is defined, use the innerButton, otherwise wrap it in a `<Link>`.

marta- replied:
It would actually make more sense if we could put the link inside the tooltip, not outside. I am aware that it was already outside originally.

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/EditButton.jsx:
```diff
@@ -49,6 +52,7 @@ EditButton.propTypes = {
   size: PropTypes.oneOf(["small", "medium", "large"]),
   className: PropTypes.string,
   admin: PropTypes.bool,
+  useEditDialog: PropTypes.bool,
```
onClick is omitted in the propTypes, please add it.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/SubjectTypeDialog.jsx:
```diff
@@ -24,16 +24,16 @@ import withStyles from '@mui/styles/withStyles';
 
-  const [ label, setLabel ] = useState(isEdit ? currentSubjectType.label : "");
+  const [ label, setLabel ] = useState("");
   const [ parentSubject, setParentSubject ] = useState(initialParent);
```
A bit out of scope, but rename to `parent`. `Subject` is incorrect anyway, (we're dealing with subject **types**), but `/SubjectTypes` is neither anyway.

```suggestion
  const [ parent, setParent ] = useState(initialParent);
```

marta- commented on modules/homepage/src/main/frontend/src/adminDashboard/AdminResourceListing.jsx:
```diff
@@ -16,23 +16,114 @@
 
 function AdminResourceListing(props) {
-  const { title, action, buttonProps, ...rest } = props;
+  const { title, columns, tableActions, buttonProps, customUrl, entryType, resourceSelectors, updateData,
```
let's rename `customUrl` (which made more sense for livetables) to `dataUrl`

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/Questionnaires.jsx:
```diff
@@ -16,57 +16,116 @@
+
+  let makeActions = ({ row }) => {
+    return (
+            <Box sx={{ display: 'flex', flexWrap: 'nowrap', float: 'right'}}>
+              <EditButton
+                entryType={entryType}
+                entryPath={row.original["@path"]}
+                admin
+              />
+              <ExportButton
+                entryPath={row.original["@path"]}
+                entryName={row.original.title}
+                entityData={row.original}
+                entryType={entryType}
+                size="medium"
+              />
+              <DeleteButton
+                entryPath={row.original["@path"]}
+                entryName={row.original.title}
+                onComplete={dialogSuccess}
+                entryType={entryType}
+                admin
+              />
```
Is there a way to generate the buttons in the `AdminResourceListing` component like we used to do with the LiveTable?

veronikaslc replied:
I'd rather not. We have custom `onClick`, `onComplete` actions tailored to each component. Also we have just 2 places of makeActions now. For development convenience it is easier to see and change what each button do in the component for now. New Admin pages don't appear that often.

veronikaslc committed:
CARDS-2618: Replace Livetable with MaterialReactTable in administration screens
Addressed codereview comments: better naming for props in use

veronikaslc committed:
CARDS-2618: Replace Livetable with MaterialReactTable in administration screens
Addressed codereview comments: refactoring EditButton code

veronikaslc commented:
> In Questionnaires the search doesn't seem to work at all

Works for me, can you please specify the exct steps you take? The search filter is defined via `customFilterFn` for Questionnaires, lines 95-99, I specifically implemented it there because you mentioned we need a search there, it looks for substrings in title and description. In general we pass custom search to `AdminResourceListing` via `customFilter` param. For example for SubjectTypes we search in path and label.

veronikaslc commented:
> In the Clinics admin screen I searched for "inpatient" and all the inpatient clinics were retained, PLUS "Canadian Patient Experience".

Yes, thats how their default global search filter works being on a greedy side. We can make our own custom filter function with exact subctring matching like for Questionnaires and SubjectTypes.

marta- commented:
> > In Questionnaires the search doesn't seem to work at all
> 
> Works for me, can you please specify the exact steps you take? The search filter is defined via `customFilterFn` for Questionnaires, lines 95-99, I specifically implemented it there because you mentioned we need a search in Questionnaires, it looks for substrings of exact match in titles and descriptions. In general we pass custom search to `AdminResourceListing` via `customFilter` param. For example for SubjectTypes we search in paths and labels.

Start in prems, go to admin/Questionnaires, type in the search box. Examples that don't work but should:
* "emerg" -> no effect
* "info" -> no effect
* "voice" -> no effect

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/Questionnaires.jsx:
```diff
@@ -16,57 +16,116 @@
-      "label": "Created by",
-      "format": "string",
+      accessorKey: "description",
+      header: "Description",
```
The description is often in markdown, let's render it properly with a `<FormattedText>`.

veronikaslc replied:
Fixed in [bbb0211](https://github.com/data-team-uhn/cards/pull/1868/commits/bbb0211393bcf5a147a8778518223adb8d68127e)

marta- commented on modules/homepage/src/main/frontend/src/adminDashboard/AdminResourceListing.jsx:
```diff
@@ -16,23 +16,114 @@
+      <MaterialReactTable
+        enableColumnActions={false}
+        enableColumnFilters={false}
+        enableSorting={false}
```
Can we enable sorting?

veronikaslc replied:
globally we can. But for columns with custom rendering manual sorting functon has to be supplied individually.

marta- replied:
I assume it's also possible to disable sorting on some columns.

The following columns should be sortable:
* Subject types: Subject Type, Subject label, Subjects, Default Order (I also really want to relabel these columns....)
* Questionnaires: title, created on
* Statistics: all except for Actions
* Clinics: all except Emergency Contact and Description

veronikaslc replied:
Found simple solution for sorting custom rendered columns, fixed in [c166227](https://github.com/data-team-uhn/cards/pull/1868/commits/c16622790ff922d168d545cae2161f6be9484143)

veronikaslc committed:
CARDS-2618: Replace Livetable with MaterialReactTable in administration screens
Addressed codereview comments: Custom filtering for Clinics page, exact match of names

veronikaslc committed:
CARDS-2618: Replace Livetable with MaterialReactTable in administration screens
Addressed codereview comments: Better exact mach filtering for all tables except custom filtering

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/SubjectTypeDialog.jsx:
```diff
@@ -59,33 +59,48 @@ function SubjectTypeDialog(props) {
     } else {
-      // if nothing changed - just move the node
       if (currentSubjectType["cards:defaultOrder"] == order &&
           currentSubjectType["cards:defaultOrder"] == order &&
```
```suggestion
```

veronikaslc replied:
Is there any suggestion? Seems the are no changes

marta- replied:
Sorry, fixed: removed duplicate line

veronikaslc committed:
CARDS-2618: Replace Livetable with MaterialReactTable in administration screens
Addressed codereview comments: Enabling sorting

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/SubjectTypes.jsx:
```diff
@@ -36,79 +36,85 @@ function getTextHierarchy (path, subjectTypes) {
-      "key": "subjectListLabel",
-      "label": "Subject list label",
-      "format": "string",
+      header: "Subject Type",
```
```suggestion
      header: "Subject type",
```

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/SubjectTypes.jsx:
```diff
@@ -36,79 +36,85 @@ function getTextHierarchy (path, subjectTypes) {
-      "key": "",
-      "label": "Subjects",
-      "format": (row) => (row.instanceCount ? <Link to={"/content.html/Subjects#" + row['@name']} title={"Show subjects of type " + row.label} underline="hover">{row.instanceCount}</Link> : "0"),
+      header: "Subject label",
```
```suggestion
      header: "Subject list label",
```

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/SubjectTypes.jsx:
```diff
@@ -36,79 +36,85 @@ function getTextHierarchy (path, subjectTypes) {
-      "key": "cards:defaultOrder",
-      "label": "Default Order",
-      "format": "string",
+      header: "Subjects",
```
```suggestion
      header: "Number of subjects",
```

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/SubjectTypes.jsx:
```diff
@@ -36,79 +36,85 @@ function getTextHierarchy (path, subjectTypes) {
-      "key": "jcr:createdBy",
-      "label": "Created by",
-      "format": "string",
+      header: "Default Order",
```
```suggestion
      header: "Order",
```

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/SubjectTypes.jsx:
```diff
@@ -36,79 +36,85 @@ function getTextHierarchy (path, subjectTypes) {
-                              onClick={() => {setIsEdit(true); setCurrentSubjectType(row); setDialogOpen(true);}}
-                            />
-                          </>),
+      header: "Pattern Hint",
```
```suggestion
      header: "Id pattern hint",
```

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/SubjectTypes.jsx:
```diff
@@ -36,79 +36,85 @@ function getTextHierarchy (path, subjectTypes) {
+      size: 20,
     },
+/*    {
+      header: "Id Pattern",
```
```suggestion
      header: "Id pattern",
```

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/SubjectTypes.jsx:
```diff
@@ -36,79 +36,85 @@ function getTextHierarchy (path, subjectTypes) {
+      accessorKey: 'cards:defaultOrder',
+      size: 20,
     },
+/*    {
+      header: "Id Pattern",
+      accessorKey: 'idPattern',
+    },*/
```
Why is id pattern commented out?

veronikaslc replied:
I was not entirely sure you would want to see this column. Should I add it?

marta- replied:
Yes please

veronikaslc replied:
Fixed in [6b28d29](https://github.com/data-team-uhn/cards/pull/1868/commits/6b28d29fd20f0b39563cdd1a6bfa383746d1889a)

veronikaslc committed:
CARDS-2618: Replace Livetable with MaterialReactTable in administration screens
Addressed codereview comments: Diplay description as formatted text

marta- commented on modules/statistics/src/main/frontend/src/Statistics/AdminStatistics.jsx:
```diff
@@ -88,14 +88,17 @@ function AdminStatistics(props) {
     },
     {
       header: "Y-axis",
+      accessorKey: "yVar",
```
Statistics sorted by yAxis ascending do not look sorted....

![image](https://github.com/user-attachments/assets/383b9d38-885f-4c0a-af03-5da41d4115e2)

veronikaslc replied:
Fixed in [016d8e0](https://github.com/data-team-uhn/cards/pull/1868/commits/016d8e0fe4420ebd66705a5cd48e112915fbb5c3)

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/SubjectTypes.jsx:
```diff
@@ -58,6 +59,7 @@ function SubjectTypes(props) {
     {
       header: "Subjects",
       size: 20,
+      accessorFn: (row) => row.instanceCount || 0,
```
Just curious why we need `original` below but not here

veronikaslc replied:
Those are by the specs from[ documentation](https://v1.material-react-table.com/docs/guides/data-columns#method-2---using-an-accessorfn-and-id)

marta- commented:
> > > In Questionnaires the search doesn't seem to work at all
> > 
> > 
> > Works for me, can you please specify the exact steps you take? The search filter is defined via `customFilterFn` for Questionnaires, lines 95-99, I specifically implemented it there because you mentioned we need a search in Questionnaires, it looks for substrings of exact match in titles and descriptions. In general we pass custom search to `AdminResourceListing` via `customFilter` param. For example for SubjectTypes we search in paths and labels.
> 
> Start in `prems`, go to admin/Questionnaires, type in the search box. Examples that don't work but should:
> 
> * "emerg" -> no effect
> * "info" -> no effect
> * "voice" -> no effect
> 
> Note: Also fails in `heracles`, works fine in `--test`.

Search is working for me in Questionnaires now.

veronikaslc committed:
CARDS-2618: Replace Livetable with MaterialReactTable in administration screens
Addressed codereview comments: Disable sorting for certain columns

veronikaslc commented:
>Search is working for me in Questionnaires now.

I guess adding the `accessorKey` worked there well. But still a question is why it worked in test mode andnot in others

veronikaslc committed:
CARDS-2618: Replace Livetable with MaterialReactTable in administration screens
Addressed codereview comments: Better column names

veronikaslc committed:
CARDS-2618: Replace Livetable with MaterialReactTable in administration screens
Addressed codereview comments: Added pattern column to the Subject types table

veronikaslc committed:
CARDS-2618: Replace Livetable with MaterialReactTable in administration screens

Removed duplicated line

Co-authored-by: Marta Girdea <651980+marta-@users.noreply.github.com>

veronikaslc committed:
CARDS-2618: Replace Livetable with MaterialReactTable in administration screens
Addressed codereview comments: Fixed typo

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/Questionnaires.jsx:
```diff
@@ -16,57 +16,122 @@
+      accessorKey: "description",
+      header: "Description",
+      enableSorting: false,
+      Cell: ({ row }) => <FormattedText>{row.original.description}</FormattedText>,
```
```suggestion
      Cell: ({ row }) => <FormattedText variant="caption">{row.original.description}</FormattedText>,
```

veronikaslc committed:
CARDS-2618: Replace Livetable with MaterialReactTable in administration screens

Better styling for table questionnaire description

Co-authored-by: Marta Girdea <651980+marta-@users.noreply.github.com>

veronikaslc committed:
CARDS-2618: Replace Livetable with MaterialReactTable in administration screens
Addressed codereview comments: Correct sorting in stat table

veronikaslc committed:
CARDS-2618: Replace Livetable with MaterialReactTable in administration screens
Removed tabs

sashaandjic reviewed (approved):
tested
QUESTION: I still need to refresh in order to see a new Subject Type. Did I miss something?

![2025-01-27_22-44-43](https://github.com/user-attachments/assets/1749bad9-40b0-4daa-8474-08a496b9a415)

veronikaslc commented:
@sashaandjic That is not video from this PR build

sashaandjic commented:
> @sashaandjic That video is not from this PR build

Oh my...so sorry.. I made a mistake by testing the wrong PR CARDS-2619.

sashaandjic reviewed (approved):
tested - love it!

sashaandjic commented:
One small thing - Can we pin actions to the right side of the grid? I got this look by removing columns:
![image](https://github.com/user-attachments/assets/fa9bb1c8-710d-4ff6-b20a-0024b5b4c0db)

veronikaslc commented:
Can not reproduce in Chrome. @sashaandjic which browser/version you use?

sashaandjic commented:
> Can not reproduce in Chrome. @sashaandjic which browser/version you use?

The problem appears in the **Statistics**.

veronikaslc committed:
CARDS-2618: Replace Livetable with MaterialReactTable in administration screens
Float stats action column to the right

veronikaslc commented:
@sashaandjic fixed

sashaandjic reviewed (approved):
tested - all good

sdumitriu commented on modules/data-entry/src/main/frontend/src/dataHomepage/SubjectTypes.jsx:
```diff
@@ -36,79 +36,89 @@ function getTextHierarchy (path, subjectTypes) {
+      size: 20,
     },
     {
-      "key": "jcr:createdBy",
```
Why are we getting rid of `created by`?

veronikaslc replied:
as per Marta's decision

sdumitriu commented on modules/patient-portal/src/main/frontend/src/patient-portal/Clinics.jsx:
```diff
@@ -41,9 +41,9 @@ function Clinics(props) {
-        format: clinicsSpecs[stat]
+        accessorKey: stat,
+        header: camelCaseToWords(stat),
+        enableSorting: stat != "emergencyContact" || stat != "description",
```
This is always true, `stat` can't be equal to both values.

veronikaslc replied:
Fixed in [ec8110d](https://github.com/data-team-uhn/cards/pull/1868/commits/ec8110db0aea80c0976ab7946ba9cca199f30f25)

sdumitriu reviewed (changes requested):
Regression: changing the parent of a subject type turns the subject type into an empty `sling:Folder`, there seems to be an extra request being sent.

veronikaslc committed:
CARDS-2618: Replace Livetable with MaterialReactTable in administration screens
Fixed logical error for enabling Clinics stat sorting column

veronikaslc committed:
CARDS-2618: Replace Livetable with MaterialReactTable in administration screens
Fixed regression in subject type parent reassigning

veronikaslc commented:
@sdumitriu Fixed regression in [2602093](https://github.com/data-team-uhn/cards/pull/1868/commits/2602093096cae6a18b9205c0ad77f5e747ee4081)
